### PR TITLE
Stop using internaldata

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Remove all traces of internaldata from splunkhecexporter.

**Testing:**
Unit tests.

**Documentation:**
N/A

### CODE REVIEW

**Comment by atoulme** [comment]
Test failures seem unrelated.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 1]
```diff
@@ -15,97 +15,134 @@
...
 	numDroppedTimeSeries := 0
-	_, numPoints := data.MetricAndDataPointCount()
-	splunkMetrics := make([]*splunk.Event, 0, numPoints)
-	for _, ocmd := range ocmds {
-		var host string
-		if ocmd.Resource != nil {
-			host = ocmd.Resource.Labels[conventions.AttributeHostHostname]
+	var splunkMetrics []*splunk.Event
+	for i := 0; i < data.ResourceMetrics().Len(); i++ {
+		rm := data.ResourceMetrics().At(i)
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:41]:
```golang
if rm.IsNil() {
  continue
}
```

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 2]
```diff
@@ -15,97 +15,134 @@
...
+			if sourceSet, isSet := rm.Resource().Attributes().Get(conventions.AttributeServiceName); isSet {
+				source = sourceSet.StringVal()
+			}
+			if sourcetypeSet, isSet := rm.Resource().Attributes().Get(splunk.SourcetypeLabel); isSet {
+				sourceType = sourcetypeSet.StringVal()
+			}
 		}
-		if host == "" {
-			host = unknownHostName
+		rm.InstrumentationLibraryMetrics().Len()
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:58]: Seems to call this without storing result or anything.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 3]
```diff
@@ -15,97 +15,134 @@
...
-					values, err := mapValues(logger, metric, tsPoint.GetValue())
-					if err != nil {
-						logger.Warn(
-							"Timeseries dropped to unexpected metric type",
-							zap.Any("metric", metric),
-							zap.Any("err", err))
-						numDroppedTimeSeries++
-						continue
+		for ilmi := 0; ilmi < rm.InstrumentationLibraryMetrics().Len(); ilmi++ {
+			ilm := rm.InstrumentationLibraryMetrics().At(ilmi)
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:66]:
```golang
if rm.IsNil() {
  continue
}
```

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 4]
```diff
@@ -15,97 +15,134 @@
...
+			}
 		}
-		if host == "" {
-			host = unknownHostName
+		rm.InstrumentationLibraryMetrics().Len()
+		commonFields := map[string]interface{}{}
+		if !rm.Resource().IsNil() {
+			rm.Resource().Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+				commonFields[k] = v.StringVal()
+			})
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:63]: Not all attributes are string, you may want to call `tracetranslator.AttributeValueToString`

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 5]
```diff
@@ -15,97 +15,134 @@
...
+			ilm := rm.InstrumentationLibraryMetrics().At(ilmi)
+			for tmi := 0; tmi < ilm.Metrics().Len(); tmi++ {
+				tm := ilm.Metrics().At(tmi)
+				switch tm.DataType() {
+				case pdata.MetricDataTypeIntGauge:
+					for gi := 0; gi < tm.IntGauge().DataPoints().Len(); gi++ {
+						fields := copy(commonFields)
+						dataPt := tm.IntGauge().DataPoints().At(gi)
+						populateLabels(fields, dataPt.LabelsMap())
+						fields[fmt.Sprintf("%s:%s", splunkMetricValue, tm.Name())] = dataPt.Value()
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:75]: This `string` can be calculated only once before iterating over all points. Also in general is faster `splunkMetricValue  + ":" + tm.Name()`

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 6]
```diff
@@ -15,97 +15,134 @@
...
-						continue
+		for ilmi := 0; ilmi < rm.InstrumentationLibraryMetrics().Len(); ilmi++ {
+			ilm := rm.InstrumentationLibraryMetrics().At(ilmi)
+			for tmi := 0; tmi < ilm.Metrics().Len(); tmi++ {
+				tm := ilm.Metrics().At(tmi)
+				switch tm.DataType() {
+				case pdata.MetricDataTypeIntGauge:
+					for gi := 0; gi < tm.IntGauge().DataPoints().Len(); gi++ {
+						fields := copy(commonFields)
+						dataPt := tm.IntGauge().DataPoints().At(gi)
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:73]: ditto for nil point. probably before copying the commonFields.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 7]
```diff
@@ -15,97 +15,134 @@
...
 func metricDataToSplunk(logger *zap.Logger, data pdata.Metrics, config *Config) ([]*splunk.Event, int, error) {
-	ocmds := internaldata.MetricsToOC(data)
 	numDroppedTimeSeries := 0
-	_, numPoints := data.MetricAndDataPointCount()
-	splunkMetrics := make([]*splunk.Event, 0, numPoints)
-	for _, ocmd := range ocmds {
-		var host string
-		if ocmd.Resource != nil {
-			host = ocmd.Resource.Labels[conventions.AttributeHostHostname]
+	var splunkMetrics []*splunk.Event
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:39]: Pre-allocate the slice:
```golang
_, dpCount := data.MetricAndDataPointCount()
splunkMetrics := make([]*splunk.Event, 0, dpCount)
```

And if you want more optimal:
```
splunkMetricsOrigs := make([]splunk.Event, dpCount)
splunkMetrics := make([]*splunk.Event, 0, dpCount)
...
  append(splunkMetrics, &splunkMetricsOrigs[i])
  // change the createEvent to fillEvent and avoid allocation.
  fillEvent(dataPt.Timestamp(), host, source, sourceType, index, fields, &splunkMetricsOrigs[i])
...
```

File: exporter/splunkhecexporter/tracedata_to_splunk.go [hunk 8]
```diff
@@ -17,43 +17,149 @@ package splunkhecexporter
...
+	Attributes map[string]interface{}  `json:"attributes,omitempty"`
+	Name       string                  `json:"name"`
+	Timestamp  pdata.TimestampUnixNano `json:"timestamp"`
+}
+
+// HecLink is a data structure holding a span link to export explicitly to Splunk HEC.
+type HecLink struct {
+	Attributes map[string]interface{} `json:"attributes,omitempty"`
+	TraceID    []byte                 `json:"trace_id"`
+	SpanID     []byte                 `json:"span_id"`
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/tracedata_to_splunk.go:36]: I think you need either base64 or base16 (hex) encoded correct? Not sure what is the right thing that you want.

Comment (bogdandrutu) [reply exporter/splunkhecexporter/tracedata_to_splunk.go:36]: So probably worth making this string and call `HexString` on the ids?

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 7 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:39]: that's probably not a good idea - in the case of a histogram, we will generate additional points.

File: exporter/splunkhecexporter/tracedata_to_splunk.go (hunk 8 above)
Comment (atoulme) [reply exporter/splunkhecexporter/tracedata_to_splunk.go:36]: yeah I thought base64 was still the way. Apparently hex is now standard, moving to that.

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 4 above)
Comment (tigrannajaryan) [reply exporter/splunkhecexporter/metricdata_to_splunk.go]: Do we need to convert non-string attribute values to string here or we want to preserve the type?

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 9]
```diff
@@ -15,97 +15,134 @@
...
-							zap.Any("err", err))
-						numDroppedTimeSeries++
-						continue
+		for ilmi := 0; ilmi < rm.InstrumentationLibraryMetrics().Len(); ilmi++ {
+			ilm := rm.InstrumentationLibraryMetrics().At(ilmi)
+			for tmi := 0; tmi < ilm.Metrics().Len(); tmi++ {
+				tm := ilm.Metrics().At(tmi)
+				switch tm.DataType() {
+				case pdata.MetricDataTypeIntGauge:
+					for gi := 0; gi < tm.IntGauge().DataPoints().Len(); gi++ {
```
Comment (tigrannajaryan) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:81]: Nit: assign `tm.IntGauge().DataPoints()` to a temp var and use. Go compiler often is unable to do this optimization.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 10]
```diff
@@ -15,97 +15,134 @@
...
-							"Timeseries dropped to unexpected metric type",
-							zap.Any("metric", metric),
-							zap.Any("err", err))
-						numDroppedTimeSeries++
-						continue
+		for ilmi := 0; ilmi < rm.InstrumentationLibraryMetrics().Len(); ilmi++ {
+			ilm := rm.InstrumentationLibraryMetrics().At(ilmi)
+			for tmi := 0; tmi < ilm.Metrics().Len(); tmi++ {
+				tm := ilm.Metrics().At(tmi)
+				switch tm.DataType() {
```
Comment (tigrannajaryan) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:79]: Check that `tm` and all other object is `!IsNill()` before accessing its properties.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 11]
```diff
@@ -114,119 +151,43 @@ func metricDataToSplunk(logger *zap.Logger, data pdata.Metrics, config *Config)
...
-	values[metricName] = summaryValue.Sum.Value
-
-	// 3. Each quantile value is converted to a gauge called <basename>_quantile
-	// and will include a dimension called quantile that specifies the quantile.
-	percentiles := summaryValue.GetSnapshot().GetPercentileValues()
-	if percentiles == nil {
-		return values, fmt.Errorf(
-			"unknown percentiles values for summary metric %q",
-			metricName)
+func timestampToEpochMilliseconds(ts pdata.TimestampUnixNano) *float64 {
```
Comment (tigrannajaryan) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:209]: Nit: the name is a bit misleading. It returns the timestamp in seconds with millisecond precision.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 12]
```diff
@@ -114,119 +151,43 @@ func metricDataToSplunk(logger *zap.Logger, data pdata.Metrics, config *Config)
...
-	splunkBounds := make([]*string, len(bounds)+1)
-	for i := 0; i < len(bounds); i++ {
-		dimValue := float64ToDimValue(bounds[i])
-		splunkBounds[i] = &dimValue
-	}
-	splunkBounds[len(splunkBounds)-1] = &infinityBoundSFxDimValue
-
-	for i, bucket := range distributionValue.Buckets {
-		values[fmt.Sprintf("%s%s", bucketMetricName, *splunkBounds[i])] = bucket.Count
+func copy(fields map[string]interface{}) map[string]interface{} {
```
Comment (tigrannajaryan) [inline exporter/splunkhecexporter/metricdata_to_splunk.go]: Nit: usually function named `copy` uses a source and destination. When the function _creates_ a copy usually it is named `clone`.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 13]
```diff
@@ -114,119 +151,43 @@ func metricDataToSplunk(logger *zap.Logger, data pdata.Metrics, config *Config)
...
+		// we want to omit the time field when sending data to the Splunk HEC so that the HEC adds a timestamp
+		// at indexing time, which will be much more useful than a 0-epoch-time value.
+		return nil
 	}
-	quantileMetricName := metricName + separator + quantileSuffix + separator
-	for _, quantile := range percentiles {
-		values[quantileMetricName+float64ToDimValue(quantile.Percentile)] = quantile.Value
 
-	}
+	val := math.Round(float64(ts)/1e6) / 1e3
```
Comment (tigrannajaryan) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:218]: Loss of precision of int64 here. Don't convert to float64 before division. float64 has only 53 bits of precision. This essentially throws away 11 bits of nanosecond precision, turning it to about 2-millisecond precision. This should avoid the problem:
```suggestion
	val := float64(ts/1e6) / 1e3
```

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 9 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go]: ok pushed that change

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 10 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:79]: added a few more checks. Will need more tests for everyone of those checks :/

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 13 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:242]: No, that makes tests fail, as 32001999345 is not rounded to 32.002 but to 32.001 in that case.
Let me try to replicate what you're saying in a test case and see if I can verify it.

Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:242]: I can't replicate. I pushed a test to show that I get a correct rounding with even a single nanosecond.

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 11 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go]: Fixed

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 10 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:82]: Fixed, and now adding tests to get to coverage threshold.

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 7 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:40]: I know that's not what the code says now, but that's the next job - mapping better to histogram for Splunk HEC.

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 12 above)
Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go]: ok changed to `cloneMap`

File: exporter/splunkhecexporter/metricdata_to_splunk.go (hunk 7 above)
Comment (bogdandrutu) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:40]: And? Still batch enough allocations, also the first suggestion should be done anyway.

Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go:40]: You're right I'll get on it.

Comment (atoulme) [reply exporter/splunkhecexporter/metricdata_to_splunk.go]: OK, I've done the first part.

**Review by bogdandrutu — APPROVED** [review]
Good starting point.

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 14]
```diff
@@ -15,97 +15,194 @@
...
+			metrics := ilm.Metrics()
+			for tmi := 0; tmi < metrics.Len(); tmi++ {
+				tm := metrics.At(tmi)
+				if tm.IsNil() {
+					continue
+				}
+				metricFieldName := splunkMetricValue + ":" + tm.Name()
+				switch tm.DataType() {
+				case pdata.MetricDataTypeIntGauge:
+					// TODO uncomment after https://github.com/open-telemetry/opentelemetry-collector/pull/2055 is merged.
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:83]: Uncomment

File: exporter/splunkhecexporter/metricdata_to_splunk.go [hunk 15]
```diff
@@ -15,97 +15,194 @@
...
+						splunkMetrics = append(splunkMetrics, sm)
 					}
-					for key, value := range values {
-						if value == nil {
-							logger.Warn(
-								"Timeseries dropped to unexpected metric type",
-								zap.Any("metric", value))
-							numDroppedTimeSeries++
+				case pdata.MetricDataTypeDoubleGauge:
+					// TODO uncomment after https://github.com/open-telemetry/opentelemetry-collector/pull/2055 is merged.
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/metricdata_to_splunk.go:101]: Uncomment.

File: exporter/splunkhecexporter/tracedata_to_splunk_test.go [hunk 16]
```diff
@@ -149,13 +149,9 @@ func makeSpan(name string, ts *pdata.TimestampUnixNano) pdata.Span {
...
 	spanLink.Attributes().InsertBool("bar", false)
 	foobarContents := pdata.NewAttributeValueArray()
-	foobarContents.InitEmpty()
-	arrContents := pdata.NewAnyValueArray()
-	arrContents.Append(pdata.NewAttributeValueString("a"))
-	arrContents.Append(pdata.NewAttributeValueString("b"))
-	foobarContents.SetArrayVal(arrContents)
+	foobarContents.ArrayVal().Append(pdata.NewAttributeValueString("a"))
+	foobarContents.ArrayVal().Append(pdata.NewAttributeValueString("b"))
 	spanLink.Attributes().Insert("foobar", foobarContents)
```
Comment (bogdandrutu) [inline exporter/splunkhecexporter/tracedata_to_splunk_test.go]: nicer I think this way :)

**Review by bogdandrutu — APPROVED** [review]

File: exporter/splunkhecexporter/tracedata_to_splunk_test.go (hunk 16 above)
Comment (atoulme) [reply exporter/splunkhecexporter/tracedata_to_splunk_test.go]: I'm just used to slapping `InitEmpty()` to any object that happens to be in reach by now.

Comment (bogdandrutu) [reply exporter/splunkhecexporter/tracedata_to_splunk_test.go]: I am working to remove that :)

**Pull request merged**